File: problems/034-capability-cut/solutions/javascript/main.js
import * as std from "std";
/** @typedef {{ to: number, rev: number, cap: bigint }} Edge */
const input = std.in.readAsString();
let scan = 0;
function nextInt() {
  while (scan < input.length && input.charCodeAt(scan) <= 32) scan++;
  let value = 0;
  while (scan < input.length) {
    const digit = input.charCodeAt(scan) - 48;
    if (digit < 0 || digit > 9) break;
    value = value * 10 + digit;
    scan++;
  }
  return value;
}
function nextBigInt() {
  while (scan < input.length && input.charCodeAt(scan) <= 32) scan++;
  const start = scan;
  while (scan < input.length) {
    const code = input.charCodeAt(scan);
    if (code < 48 || code > 57) break;
    scan++;
  }
  return BigInt(input.slice(start, scan));
}
let n = nextInt(),
  m = nextInt(),
  sn = nextInt(),
  tn = nextInt();
/** @type {bigint[]} */
const c = Array.from({ length: n }, () => nextBigInt());
/** @type {number[]} */
const en = Array.from({ length: sn }, () => nextInt() - 1);
/** @type {number[]} */
const dn = Array.from({ length: tn }, () => nextInt() - 1);
const V = 2 * n + 2,
  S = 2 * n,
  T = S + 1;
/** @type {Edge[][]} */
const g = Array.from({ length: V }, () => []);
/**
 * @param {number} u
 * @param {number} v
 * @param {bigint} z
 */
const add = (u, v, z) => {
  g[u].push({ to: v, rev: g[v].length, cap: z });
  g[v].push({ to: u, rev: g[u].length - 1, cap: 0n });
};
const inf = c.reduce((x, y) => x + y, 1n);
for (let i = 0; i < n; i++) add(2 * i, 2 * i + 1, c[i]);
while (m--) {
  const u = nextInt() - 1, v = nextInt() - 1;
  add(2 * u + 1, 2 * v, inf);
}
for (const u of en) add(S, 2 * u, inf);
for (const u of dn) add(2 * u + 1, T, inf);
let flow = 0n;
for (;;) {
  const level = new Int32Array(V).fill(-1), q = [S];
  level[S] = 0;
  for (let h = 0; h < q.length; h++) {
    const u = q[h];
    for (const e of g[u]) {
      if (e.cap > 0n && level[e.to] < 0) {
        level[e.to] = level[u] + 1;
        q.push(e.to);
      }
    }
  }
  if (level[T] < 0) break;
  const it = new Int32Array(V);
  /**
   * @param {number} u
   * @param {bigint} f
   * @returns {bigint}
   */
  const dfs = (u, f) => {
    if (u === T) return f;
    while (it[u] < g[u].length) {
      const e = g[u][it[u]];
      if (e.cap > 0n && level[e.to] === level[u] + 1) {
        const z = dfs(e.to, f < e.cap ? f : e.cap);
        if (z > 0n) {
          e.cap -= z;
          g[e.to][e.rev].cap += z;
          return z;
        }
      }
      it[u]++;
    }
    return 0n;
  };
  for (;;) {
    const z = dfs(S, inf);
    if (z === 0n) break;
    flow += z;
  }
}
std.out.puts(`COST ${flow}\n`);

File: problems/034-capability-cut/solutions/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
typedef unsigned long long U;
static int *head, *to, *next, *level, *it, ec, T;
static U *cap;
void add(int u, int v, U c) {
  to[ec] = v;
  cap[ec] = c;
  next[ec] = head[u];
  head[u] = ec++;
  to[ec] = u;
  cap[ec] = 0;
  next[ec] = head[v];
  head[v] = ec++;
}
U dfs(int u, U f) {
  if (u == T)
    return f;
  for (int *e = &it[u]; *e >= 0; *e = next[*e])
    if (cap[*e] && level[to[*e]] == level[u] + 1) {
      U z = dfs(to[*e], f < cap[*e] ? f : cap[*e]);
      if (z) {
        cap[*e] -= z;
        cap[*e ^ 1] += z;
        return z;
      }
    }
  return 0;
}
int main(void) {
  int n, m, sn, tn;
  if (scanf("%d%d%d%d", &n, &m, &sn, &tn) != 4)
    return 0;
  U *c = malloc(sizeof(U) * (size_t)n), sum = 0;
  for (int i = 0; i < n; i++) {
    scanf("%llu", &c[i]);
    sum += c[i];
  }
  int *en = malloc(sizeof(int) * (size_t)sn),
      *dn = malloc(sizeof(int) * (size_t)tn);
  for (int i = 0; i < sn; i++) {
    scanf("%d", &en[i]);
    en[i]--;
  }
  for (int i = 0; i < tn; i++) {
    scanf("%d", &dn[i]);
    dn[i]--;
  }
  int *eu = malloc(sizeof(int) * (size_t)(m ? m : 1)),
      *ev = malloc(sizeof(int) * (size_t)(m ? m : 1));
  for (int i = 0; i < m; i++) {
    scanf("%d%d", &eu[i], &ev[i]);
    eu[i]--;
    ev[i]--;
  }
  int V = 2 * n + 2, S = 2 * n, lim = 2 * (n + m + sn + tn) + 2;
  T = S + 1;
  head = malloc(sizeof(int) * (size_t)V);
  to = malloc(sizeof(int) * (size_t)lim);
  next = malloc(sizeof(int) * (size_t)lim);
  cap = malloc(sizeof(U) * (size_t)lim);
  level = malloc(sizeof(int) * (size_t)V);
  it = malloc(sizeof(int) * (size_t)V);
  memset(head, 255, sizeof(int) * (size_t)V);
  U inf = sum + 1;
  for (int i = 0; i < n; i++)
    add(2 * i, 2 * i + 1, c[i]);
  for (int i = 0; i < m; i++)
    add(2 * eu[i] + 1, 2 * ev[i], inf);
  for (int i = 0; i < sn; i++)
    add(S, 2 * en[i], inf);
  for (int i = 0; i < tn; i++)
    add(2 * dn[i] + 1, T, inf);
  int *q = malloc(sizeof(int) * (size_t)V);
  U flow = 0;
  for (;;) {
    memset(level, 255, sizeof(int) * (size_t)V);
    int l = 0, r = 0;
    level[S] = 0;
    q[r++] = S;
    while (l < r) {
      int u = q[l++];
      for (int e = head[u]; e >= 0; e = next[e])
        if (cap[e] && level[to[e]] < 0) {
          level[to[e]] = level[u] + 1;
          q[r++] = to[e];
        }
    }
    if (level[T] < 0)
      break;
    memcpy(it, head, sizeof(int) * (size_t)V);
    U z;
    while ((z = dfs(S, inf)) != 0)
      flow += z;
  }
  printf("COST %llu\n", flow);
  return 0;
}
